### PR TITLE
Move bio_tags_to_spans out of SpanBasedF1Metric

### PR DESCRIPTION
**Pull request by DeNeutoy**
This method will be useful generally (for span based models which use the Ontonotes data in particular), so it was worth shifting it out of `SpanBasedF1Measure`.

### CODE REVIEW

File: allennlp/data/dataset_readers/dataset_utils/ontonotes.py [hunk 1]
```diff
@@ -429,3 +430,81 @@ def _process_span_annotations_for_word(annotations: List[str],
...
+    ----------
+    tag_sequence : List[int], required.
+        The integer class labels for a sequence.
+    classes_to_ignore : List[str], optional (default = None).
+        A list of string class labels `excluding` the bio tag
+        which should be ignored when extracting spans.
+
+    Returns
+    -------
+    spans : Set[Tuple[Tuple[int, int], str]]
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/dataset_utils/ontonotes.py]: This doesn't match the return type you've given above.

File: allennlp/data/dataset_readers/dataset_utils/ontonotes.py [hunk 2]
```diff
@@ -429,3 +430,81 @@ def _process_span_annotations_for_word(annotations: List[str],
...
+    """
+    Given a sequence corresponding to BIO tags, extracts spans.
+    Spans are inclusive and can be of zero length, representing a single word span.
+    Ill-formed spans are also included (i.e those which do not start with a "B-LABEL"),
+    as otherwise it is possible to get a perfect precision score whilst still predicting
+    ill-formed spans in addition to the correct spans.
+
+    Parameters
+    ----------
+    tag_sequence : List[int], required.
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/dataset_utils/ontonotes.py]: This type doesn't match the input type above.

File: allennlp/data/dataset_readers/dataset_utils/ontonotes.py [hunk 3]
```diff
@@ -429,3 +430,81 @@ def _process_span_annotations_for_word(annotations: List[str],
             # Exiting a span, so we reset the current span label for this annotation.
             if ")" in annotation:
                 current_span_labels[annotation_index] = None
+
+
+def bio_tags_to_spans(tag_sequence: List[str],
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/dataset_utils/ontonotes.py]: Will we want this outside of ontonotes?  I'm just wondering if this is the best place for this code to live.

File: allennlp/training/metrics/span_based_f1_measure.py [hunk 4]
```diff
@@ -111,91 +112,23 @@ def __call__(self,
...
                 # completely padded. These contribute nothing, so we skip these rows.
                 continue
-            prediction_spans = self._extract_spans(sequence_prediction[:length].tolist())
-            gold_spans = self._extract_spans(sequence_gold_label[:length].tolist())
+
+            predicted_string_labels = [self._label_vocabulary[label_id]
+                                       for label_id in sequence_prediction[:length].tolist()]
+            gold_string_labels = [self._label_vocabulary[label_id]
+                                  for label_id in sequence_gold_label[:length].tolist()]
+            prediction_spans = bio_tags_to_spans(predicted_string_labels, self._ignore_classes)
```
Comment (matt-gardner) [inline allennlp/training/metrics/span_based_f1_measure.py]: Above you have `predicted` and here you have `prediction` - can you be consistent between those two?  No strong preference on which one you should use.

File: allennlp/data/dataset_readers/dataset_utils/ontonotes.py (hunk 3 above)
Comment (DeNeutoy) [reply allennlp/data/dataset_readers/dataset_utils/ontonotes.py]: We can move it out once it starts to be used elsewhere?

Comment (matt-gardner) [reply allennlp/data/dataset_readers/dataset_utils/ontonotes.py]: Sure, seems reasonable.

**Pull request merged**